File: app/src/main/java/com/hfad/avc/managers/DBManager.java
package com.hfad.avc.managers;

import android.annotation.SuppressLint;
import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.provider.ContactsContract;
import android.util.Log;

import androidx.work.PeriodicWorkRequest;

import com.hfad.avc.R;
import com.hfad.avc.dagger.ComponentManager;
import com.hfad.avc.data.database.AppDatabase;
import com.hfad.avc.data.database.ContactDao;
import com.hfad.avc.data.model.Contact;
import com.hfad.avc.data.model.Template;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.rxjava3.annotations.NonNull;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;

public class DBManager {
   private PeriodicWorkRequest myWorkRequest;
   private static final String TAG = "DBManager";
   @Inject
   AppDatabase db;
   private List<Contact> contactsList;
   private List<Template> templateList;
   @Inject
   Context context;
   private final Contact contact;
   private final Template template;

   @Singleton
   public DBManager() {
      ComponentManager.Companion.getInstance().appComponent.inject(this);
      contact = new Contact();
      template = new Template();
   }

   public List<Contact> getContactList(int position) {
      ContactDao db_contactDao = db.contactDao();
      try {
         switch (position) {
            case 0:
               return this.contactsList = db_contactDao.all();
            case 1:
               //return this.contactsList = db_contactDao.getAllCongratulations();
            case 2:
               return this.contactsList = db_contactDao.getFavoriteContact();
         }
      } catch (Exception e) {
         Log.e(TAG, "getContactList: " + e);
      }
      return this.contactsList;
   }

   public List<Template> getTemplateList(int position) {
      try {
         switch (position) {
            case 0:
               return this.templateList = db.templateDao().getAll();
            case 1:
               return this.templateList = db.templateDao().getFavorite();
         }
      } catch (Exception e) {
         Log.e(TAG, "getContactList: " + e);
      }
      return this.templateList;
   }

   @SuppressWarnings({"unused", "RedundantSuppression"})
//   public Completable createWorkRequest(Data data, long seconds, String id) {
//      return Observable.just(data)
//            .flatMapCompletable(data1 -> {
//               this.myWorkRequest = new PeriodicWorkRequest.Builder(WorkerManager.class, seconds, TimeUnit.MINUTES)
//                     .setInputData(data1)
//                     .setInitialDelay(10, TimeUnit.SECONDS)
//                     .build();
//               WorkManager.getInstance(context).enqueue(this.myWorkRequest);
//               return Completable.complete();
//            }).doFinally(() -> {
//               updateListIDWorker(id);
//            });
//   }
//
//   public Completable cancelWorkRequest(String id) {
//      Contact contact = db.contactDao().getById(id);
//      ArrayList<String> listIDWorker = new ArrayList<>(contact.getListIDWorker());
//      return Observable.just(listIDWorker)
//            .flatMapCompletable(listID -> {
//               for (String idfromList : listID) {
//                  UUID uuid = UUID.fromString(idfromList);
//                  WorkManager.getInstance(context).cancelWorkById(uuid);
//               }
//               return Completable.complete();
//            }).doFinally(() -> {
//               clineListIDWorker(id);
//            });
//   }

//   public void updateListIDWorker(String id) {
//      //FIXME добавляет но медленно, лучше в RX
//      Contact contact = db.contactDao().getById(id);
//      ArrayList<String> listIDWorker = new ArrayList<>(contact.getListIDWorker());
//      listIDWorker.add(this.myWorkRequest.getId().toString());
//      contact.setListIDWorker(listIDWorker);
//      db.contactDao().update(contact);
//   }
//
//   public void clineListIDWorker(String id) {
//      //FIXME удаляет но медленно, лучше в RX
//      Contact contact = db.contactDao().getById(id);
//      contact.setListIDWorker(Collections.emptyList());
//      db.contactDao().update(contact);
//   }

   public Observable<List<Contact>> loadDB() {
      ContentResolver contentResolver = context.getContentResolver();
      @SuppressLint("Recycle") Cursor cursor = contentResolver.query(
            ContactsContract.Contacts.CONTENT_URI,
            null,
            null,
            null,
            null
      );
      try {
         List<Contact> contactsListForUpdate = db.contactDao().all();
         Map<String, Contact> map = new HashMap<>();
         if (cursor != null) {
            while (cursor.moveToNext()) {
               @SuppressLint("Range") String id = cursor.getString(cursor.getColumnIndex(ContactsContract.Contacts._ID));
               contact.setId(id);
               @SuppressLint("Range") String name = cursor.getString(cursor.getColumnIndex(
                     ContactsContract.Contacts.DISPLAY_NAME));
               contact.setName(name != null ? name : "");
               @SuppressLint("Range") String uriThumbnail = cursor.getString(cursor.getColumnIndex(
                     ContactsContract.Contacts.PHOTO_THUMBNAIL_URI));
               contact.setUriThumbnail(uriThumbnail != null ? uriThumbnail : "");
               @SuppressLint("Range") String uriFull = cursor.getString(cursor.getColumnIndex(
                     ContactsContract.Contacts.PHOTO_URI));
               contact.setUriFull(uriFull != null ? uriFull : "");
               @SuppressLint("Range") String has_phone = cursor.getString(cursor.getColumnIndex(
                     ContactsContract.Contacts.HAS_PHONE_NUMBER));
               if (Integer.parseInt(has_phone) > 0) {
                  Cursor pCur;
                  pCur = context.getContentResolver().query(
                        ContactsContract.CommonDataKinds.Phone.CONTENT_URI,
                        null,
                        ContactsContract.CommonDataKinds.Phone.CONTACT_ID + " = ?",
                        new String[]{id},
                        null);
                  while (pCur.moveToNext()) {
                     @SuppressLint("Range") String phone = pCur.getString(pCur.getColumnIndex(
                           ContactsContract.CommonDataKinds.Phone.NUMBER));
                     contact.addToPhoneList(phone);
                  }
                  pCur.close();
               }
               db.contactDao().upsert(contact);
               contact.clear();
            }
            String[] countries = context.getResources().getStringArray(R.array.congratulations_templates);
            if (db.templateDao().getSize().equals("0")) {
               for (int i = 0; i < countries.length; i++) {
                  this.template.setId(String.valueOf(i));
                  template.setTextTemplate(countries[i]);
                  template.setFavorite(false);
                  db.templateDao().insert(template);
               }
            }
            return Observable.just(db.contactDao().all());
         }
      } catch (Exception exception) {
         Log.e(TAG, "loadDB: exception: ", exception);
      }
      return Observable.just(Collections.emptyList());
   }

   public @NonNull Single<List<Contact>> Filt(String searchText) {
      contactsList = getContactList(0);
      return Observable.fromIterable(this.contactsList)
            .filter(contact -> {
               String searchTextfinal = searchText.toLowerCase();
               if (contact.getName().toLowerCase().contains(searchTextfinal)) {
                  return true;
               }
               return contact.getPhoneList().get(0).toLowerCase().contains(searchTextfinal);
            })
            .toList();

   }
}

File: app/src/main/java/com/lastaurus/automatic_congratulations/managers/DBManager.kt
package com.lastaurus.automatic_congratulations.managers

import android.annotation.SuppressLint
import android.content.Context
import android.database.Cursor
import android.provider.ContactsContract
import android.util.Log
import androidx.work.Data
import androidx.work.PeriodicWorkRequest
import androidx.work.WorkManager
import com.lastaurus.automatic_congratulations.R
import com.lastaurus.automatic_congratulations.dagger.ComponentManager.Companion.instance
import com.lastaurus.automatic_congratulations.data.database.AppDatabase
import com.lastaurus.automatic_congratulations.data.model.Contact
import com.lastaurus.automatic_congratulations.data.model.Template
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import java.util.*
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton


class DBManager @Singleton constructor() {

   @Inject
   lateinit var db: AppDatabase

   @Inject
   lateinit var context: Context
   private val contact: Contact by lazy { Contact() }
   private val template: Template by lazy { Template() }
   private val scope: CoroutineScope by lazy { CoroutineScope(Job()) }
   private var dbJob: Job? = null

   init {
      instance.appComponent.inject(this)
      loadTemplateList()
   }

   fun cancelWorkRequest(idCongratulation: Int) {
      scope.launch {
         db.congratulationsDao().getById(idCongratulation)?.let { congratulation ->
            congratulation.getIdWorker()?.let {
               WorkManager.getInstance(context).cancelWorkById(it)
               congratulation.claenIdWorker()
            }
         }
      }
   }


   fun setEveryYearInterval(idCongratulation: Int) {
      scope.launch {
         val congratulationTime = db.congratulationsDao().getById(idCongratulation)?.getDateTime()
         val currentTime = Calendar.getInstance().timeInMillis
         var seconds = (congratulationTime?.minus(currentTime))
         seconds = seconds?.div(1000)
         seconds?.let {
            if (seconds < 0) {
               seconds += 31536000
            }
         }
         //TODO
      }
   }

   fun createWorkRequest(data: Data, idCongratulation: Int) {
      var periodicWorkRequest: PeriodicWorkRequest?
      scope.launch {
         periodicWorkRequest =
            PeriodicWorkRequest.Builder(
               WorkerManager::class.java,
               365,
               TimeUnit.DAYS,
               24,
               TimeUnit.HOURS,
            )
               .setInputData(data)
               .setInitialDelay(10, TimeUnit.SECONDS)
               .build()

         periodicWorkRequest?.let {
            db.congratulationsDao().getById(idCongratulation)?.let { congratulation ->
               congratulation.getIdWorker()?.let { uuid ->
                  WorkManager.getInstance(context).cancelWorkById(uuid)
               }
               congratulation.setIdWorker(it.id)
               db.congratulationsDao().update(congratulation)
            }
            WorkManager.getInstance(context).enqueue(it)
         }
      }
   }

   fun loadSistemContactList() {
      dbJob = scope.launch {
         val contentResolver = context.contentResolver
         @SuppressLint("Recycle") val cursor = contentResolver.query(
            ContactsContract.Contacts.CONTENT_URI,
            null,
            null,
            null,
            null
         )
         try {
            cursor?.let {
               while (cursor.moveToNext()) {
                  @SuppressLint("Range") val id =
                     cursor.getString(cursor.getColumnIndex(ContactsContract.Contacts._ID))
                  contact.setId(db.contactDao().size)
                  @SuppressLint("Range") val name = cursor.getString(
                     cursor.getColumnIndex(
                        ContactsContract.Contacts.DISPLAY_NAME
                     )
                  )
                  contact.setName(name ?: "")
                  @SuppressLint("Range") val uriThumbnail = cursor.getString(
                     cursor.getColumnIndex(
                        ContactsContract.Contacts.PHOTO_THUMBNAIL_URI
                     )
                  )
                  contact.setUriThumbnail(uriThumbnail ?: "")
                  @SuppressLint("Range") val uriFull = cursor.getString(
                     cursor.getColumnIndex(
                        ContactsContract.Contacts.PHOTO_URI
                     )
                  )
                  contact.setUriFull(uriFull ?: "")
                  @SuppressLint("Range") val has_phone = cursor.getString(
                     cursor.getColumnIndex(
                        ContactsContract.Contacts.HAS_PHONE_NUMBER
                     )
                  )
                  if (has_phone.toInt() > 0) {
                     val pCur: Cursor? = context.contentResolver.query(
                        ContactsContract.CommonDataKinds.Phone.CONTENT_URI,
                        null,
                        ContactsContract.CommonDataKinds.Phone.CONTACT_ID + " = ?", arrayOf(id),
                        null
                     )
                     while (pCur!!.moveToNext()) {
                        @SuppressLint("Range") val phone = pCur.getString(
                           pCur.getColumnIndex(
                              ContactsContract.CommonDataKinds.Phone.NUMBER
                           )
                        )
                        contact.addToPhoneList(phone)
                     }
                     pCur.close()
                  }
                  db.contactDao().insert(contact)
                  db.contactDao().update(contact)
                  contact.clear()
               }
            }
         } catch (exception: Exception) {
            Log.e("gera", "loadSistemContactList", exception)
         }
      }
   }

   fun loadTemplateList() {
      scope.launch {
         val countries = context.resources.getStringArray(R.array.congratulations_templates)
         if (db.templateDao().size == 0) {
            for (i in countries.indices) {
               template.setId(i)
               template.setTextTemplate(countries[i])
               template.setFavorite(false)
               db.templateDao().insert(template)
            }
         }
      }
   }

   fun cancelJob() {
      dbJob?.cancel()
   }
}